aten/src/ATen/native/Math.h

```diff
@@ -2177,4 +2177,417 @@ static inline C10_HOST_DEVICE T calc_log_ndtr(T x) {
+    };
+
+    static const T Q[] = {
+            -2.31581873324120129819e-5,
+            +5.39605580493303397842e-4,
+            -4.45641913851797240494e-3,
+            +1.18139785222060435552e-2,
+            +3.58236398605498653373e-2,
+            -2.34591795718243348568e-1,
+            +7.14304917030273074085e-2,
+            +1.00000000000000000320e+0,
+    };
+
+    int sign_gamma = 1;
+
+    if (!std::isfinite(x)) return x;
+
+    T p;
+    T q = std::abs(x);
+    T z;
+
+    if (q > T(33.0)) {
+        if (x < T(0.0)) {
+            p = std::floor(q);
+
+            if (p == q) return std::numeric_limits<T>::infinity();
+
+            int i = p;
+
+            if ((i & 1) == 0) sign_gamma = -1;
+
+            z = q - p;
+
+            if (z > T(0.5)) {
+                p = p + T(1.0);
+                z = q - p;
+            }
+
+            z = q * std::sin(c10::pi<T> * z);
+
+            if (z == T(0.0)) return sign_gamma * std::numeric_limits<T>::infinity();
+
+            z = std::abs(z);
+
+            z = c10::pi<T> / (z * stirling_gamma_approximation(q));
+        } else {
+            z = stirling_gamma_approximation(x);
+        }
+
+        return sign_gamma * z;
+    }
+
+    z = T(1.0);
+
+    while (x >= T(3.0)) {
+        x = x - T(1.0);
+        z = z * x;
+    }
+
+    while (x < T(0.0)) {
+        if (x > T(-0.000000001)) {
+            if (x == T(0.0)) return std::numeric_limits<T>::infinity();
+
+            return z / ((T(1.0) + T(0.5772156649015329) * x) * x);
+        }
+
+        z = z / x;
+        x = x + T(1.0);
+    }
+
+    while (x < T(2.0)) {
+        if (x < T(0.000000001)) {
+            if (x == T(0.0)) return std::numeric_limits<T>::infinity();
+
+            return z / ((T(1.0) + T(0.5772156649015329) * x) * x);
+        }
+
+        z = z / x;
+        x = x + T(1.0);
+    }
+
+    if (x == T(2.0)) return z;
+
+    x = x - T(2.0);
+
+    p = 0.0;
+    q = 0.0;
+
+    for (int index = 0; index <= 6; index++) p = p * x + P[index];
+    for (int index = 0; index <= 7; index++) q = q * x + Q[index];
+
+    return z * p / q;
+}
+
+template<typename T>
+T ln_beta_asymptotic_expansion(T x, T y, int *sign) {
+    return reentrant_ln_gamma(y, sign) - (y * std::log(x)) + (y * (T(1.0) - y) / (T(2.0) * x)) + (y * (T(1.0) - y) * (T(1.0) - T(2.0) * y) / (T(12.0) * x * x)) + (-y * y * (T(1.0) - y) * (T(1.0) - y) / (T(12.0) * x * x * x));
+}
+
+template<typename T>
+T ln_beta_forward(T x, T y) {
+    int sign = 1;
+
+    if (x <= T(0.0)) {
+        if (x == std::floor(x)) {
+            if (x == int(x)) return ln_beta_negative(int(x), y);
+
+            return sign * std::numeric_limits<T>::infinity();
+        }
+    }
+
+    if (y <= T(0.0)) {
+        if (y == std::floor(y)) {
+            if (y == int(y)) return ln_beta_negative(int(y), x);
+
+            return sign * std::numeric_limits<T>::infinity();
+        }
+    }
+
+    T z;
+
+    if (std::abs(x) < std::abs(y)) {
+        z = x;
```

Comment:
nit: this seems fine but conceptually it seems more like

```
std::swap(x, y)
```

because the value of `z` isn't used before it's overwritten below, I think